ckanext/xroad_integration/logic/action.py

```diff
@@ -333,6 +340,35 @@ def _convert_xroad_datetime_list_to_datetime(value):
         return None
 
 
+def check_start_and_end_date(start_date, end_date):
+    if end_date and not start_date:
+        raise ValueError("Please give start date to go with the end date")
+
+    # If request includes start_date use that (and verify it is a valid date)
+    # else default to yesterday
+    if start_date:
+        start_date = datetime.datetime.strptime(start_date, '%Y-%m-%d')
+        if start_date > datetime.datetime.now():
+            raise ValueError("Start date cannot be in the future")
+    else:
+        start_date = datetime.datetime.now() - relativedelta.relativedelta(days=+1)
+
+    # If request includes end_date use that (and verify it is a valid date)
+    # else default to current time
+    if end_date:
+        end_date = datetime.datetime.strptime(end_date, '%Y-%m-%d')
+        if start_date > end_date:
+            raise ValueError("Start date cannot be later than end date")
+        if end_date > datetime.datetime.now():
+            raise ValueError("End date cannot be in the future")
+    else:
+        end_date = datetime.datetime.now()
+
+    start_date = datetime.datetime.strftime(start_date, "%Y-%m-%d")
+    end_date = datetime.datetime.strftime(end_date, "%Y-%m-%d")
```

Comment:
Here we seem to have a function for much of what is needed above, but it also contains default values.

I'm thinking if this should be split into two functions 1. parse a date (return None if the string is empty/None, raise an exception if it is invalid), 2. validate the interval after using `or` to apply default values. At the call site these could all be parametrized as needed and exceptions would be used to handle error cases. Something like

```yesterday = datetime.datetime.now() - relativedelta.relativedelta(days=+1)
try:
  start_date, end_date = validate_interval(parse_date(start_date_str) or yesterday, parse_date(end_date_str))
except ValueError as e:
  # handle date format error OR interval error
  click.fail(e.args[0])```